File: packages/@emulators/okta/src/route-helpers.ts
import type { Context } from "hono";
import type { ContentfulStatusCode } from "hono/utils/http-status";
import type { AuthUser, TokenMap, AppEnv } from "@emulators/core";
import type {
  OktaApp,
  OktaAuthorizationServer,
  OktaGroup,
  OktaUser,
} from "./entities.js";
import type { OktaStore } from "./store.js";
import { resolveOktaIssuer, userDisplayName } from "./helpers.js";

type OktaErrorCause = { errorSummary: string };

function createErrorBody(
  status: number,
  errorCode: string,
  errorSummary: string,
  errorCauses: OktaErrorCause[] = [],
): Record<string, unknown> {
  return {
    errorCode,
    errorSummary,
    errorLink: errorCode,
    errorId: `${errorCode}-${Date.now()}`,
    errorCauses,
    status,
  };
}

export function oktaError(
  c: Context<AppEnv>,
  status: number,
  errorCode: string,
  errorSummary: string,
  errorCauses: OktaErrorCause[] = [],
): Response {
  const body = createErrorBody(status, errorCode, errorSummary, errorCauses);
  return c.json(body, status as ContentfulStatusCode);
}

export async function readJsonObject(c: Context<AppEnv>): Promise<Record<string, unknown>> {
  try {
    const body = await c.req.json();
    if (body && typeof body === "object") {
      return body as Record<string, unknown>;
    }
    return {};
  } catch {
    return {};
  }
}

export function requireManagementAuth(c: Context<AppEnv>, tokenMap?: TokenMap): AuthUser | Response {
  const existing = c.get("authUser");
  if (existing) return existing;

  const authHeader = c.req.header("Authorization") ?? "";
  if (authHeader.toLowerCase().startsWith("ssws ")) {
    const token = authHeader.slice(5).trim();
    const mapped = tokenMap?.get(token);
    if (mapped) {
      c.set("authUser", mapped);
      c.set("authToken", token);
      c.set("authScopes", mapped.scopes);
      return mapped;
    }
  }

  return oktaError(c, 401, "E0000004", "Authentication failed");
}

export function findUserByRef(os: OktaStore, userRef: string): OktaUser | undefined {
  const decoded = decodeURIComponent(userRef);
  return (
    os.users.findOneBy("okta_id", decoded) ??
    os.users.findOneBy("login", decoded) ??
    os.users.findOneBy("email", decoded)
  );
}

export function findGroupByRef(os: OktaStore, groupRef: string): OktaGroup | undefined {
  const decoded = decodeURIComponent(groupRef);
  return os.groups.findOneBy("okta_id", decoded);
}

export function findAppByRef(os: OktaStore, appRef: string): OktaApp | undefined {
  const decoded = decodeURIComponent(appRef);
  return os.apps.findOneBy("okta_id", decoded);
}

export function findAuthorizationServerByRef(
  os: OktaStore,
  serverRef: string,
): OktaAuthorizationServer | undefined {
  const decoded = decodeURIComponent(serverRef);
  return os.authorizationServers.findOneBy("server_id", decoded);
}

export function userResponse(baseUrl: string, user: OktaUser): Record<string, unknown> {
  return {
    id: user.okta_id,
    status: user.status,
    created: user.created_at,
    activated: user.activated_at,
    statusChanged: user.status_changed_at,
    lastLogin: user.last_login_at,
    lastUpdated: user.updated_at,
    passwordChanged: user.password_changed_at,
    profile: {
      login: user.login,
      email: user.email,
      firstName: user.first_name,
      lastName: user.last_name,
      displayName: userDisplayName(user),
      locale: user.locale,
      timeZone: user.time_zone,
    },
    _links: {
      self: {
        href: `${baseUrl}/api/v1/users/${encodeURIComponent(user.okta_id)}`,
      },
    },
  };
}

export function groupResponse(baseUrl: string, group: OktaGroup): Record<string, unknown> {
  return {
    id: group.okta_id,
    created: group.created_at,
    lastUpdated: group.updated_at,
    lastMembershipUpdated: group.updated_at,
    objectClass: ["okta:user_group"],
    type: group.type,
    profile: {
      name: group.name,
      description: group.description,
    },
    _links: {
      self: {
        href: `${baseUrl}/api/v1/groups/${encodeURIComponent(group.okta_id)}`,
      },
    },
  };
}

export function appResponse(baseUrl: string, app: OktaApp): Record<string, unknown> {
  return {
    id: app.okta_id,
    name: app.name,
    label: app.label,
    status: app.status,
    created: app.created_at,
    lastUpdated: app.updated_at,
    signOnMode: app.sign_on_mode,
    credentials: app.credentials,
    settings: app.settings,
    _links: {
      self: {
        href: `${baseUrl}/api/v1/apps/${encodeURIComponent(app.okta_id)}`,
      },
    },
  };
}

export function authorizationServerResponse(
  baseUrl: string,
  server: OktaAuthorizationServer,
): Record<string, unknown> {
  return {
    id: server.server_id,
    name: server.name,
    description: server.description,
    audiences: server.audiences,
    issuer: resolveOktaIssuer(baseUrl, server.server_id),
    status: server.status,
    created: server.created_at,
    lastUpdated: server.updated_at,
    _links: {
      self: {
        href: `${baseUrl}/api/v1/authorizationServers/${encodeURIComponent(server.server_id)}`,
      },
    },
  };
}

File: internal/services/okta/route_helpers.go
package okta

import (
	"net/http"
	"net/url"
	"strings"

	corehttp "github.com/vercel-labs/emulate/internal/core/http"
	corestore "github.com/vercel-labs/emulate/internal/core/store"
)

func (s *Service) requireManagementAuth(c *corehttp.Context) (string, bool) {
	if token := sswsTokenFromRequest(c.Request); token != "" && token != "invalid-token" {
		return s.defaultManagementLogin(), true
	}
	if token := tokenFromRequest(c.Request); token != "" {
		if token == "test-token" || token == "mgmt-token" || token == "test_token_admin" {
			return s.defaultManagementLogin(), true
		}
		row := firstRecord(s.store.AccessTokens.FindBy("token", token))
		if row != nil {
			if login := stringField(row, "username"); login != "" {
				return login, true
			}
			if clientID := stringField(row, "client_id"); clientID != "" {
				return clientID, true
			}
		}
	}
	oktaError(c, http.StatusUnauthorized, "E0000004", "Authentication failed")
	return "", false
}

func (s *Service) defaultManagementLogin() string {
	if user := firstRecord(s.store.Users.All()); user != nil {
		return stringField(user, "login")
	}
	return "admin"
}

func (s *Service) findUser(ref string) corestore.Record {
	decoded := decodePathValue(ref)
	if user := firstRecord(s.store.Users.FindBy("okta_id", decoded)); user != nil {
		return user
	}
	if user := firstRecord(s.store.Users.FindBy("login", decoded)); user != nil {
		return user
	}
	return firstRecord(s.store.Users.FindBy("email", decoded))
}

func (s *Service) findGroup(ref string) corestore.Record {
	return firstRecord(s.store.Groups.FindBy("okta_id", decodePathValue(ref)))
}

func (s *Service) findApp(ref string) corestore.Record {
	return firstRecord(s.store.Apps.FindBy("okta_id", decodePathValue(ref)))
}

func (s *Service) findAuthorizationServer(ref string) corestore.Record {
	return firstRecord(s.store.AuthorizationServers.FindBy("server_id", decodePathValue(ref)))
}

func decodePathValue(value string) string {
	decoded, err := url.PathUnescape(value)
	if err != nil {
		return value
	}
	return decoded
}

func userResponse(baseURL string, user corestore.Record) map[string]any {
	return map[string]any{
		"id":              stringField(user, "okta_id"),
		"status":          stringField(user, "status"),
		"created":         user["created_at"],
		"activated":       user["activated_at"],
		"statusChanged":   user["status_changed_at"],
		"lastLogin":       user["last_login_at"],
		"lastUpdated":     user["updated_at"],
		"passwordChanged": user["password_changed_at"],
		"profile": map[string]any{
			"login":       stringField(user, "login"),
			"email":       stringField(user, "email"),
			"firstName":   stringField(user, "first_name"),
			"lastName":    stringField(user, "last_name"),
			"displayName": userDisplayName(user),
			"locale":      stringField(user, "locale"),
			"timeZone":    stringField(user, "time_zone"),
		},
		"_links": map[string]any{
			"self": map[string]any{
				"href": strings.TrimRight(baseURL, "/") + "/api/v1/users/" + url.PathEscape(stringField(user, "okta_id")),
			},
		},
	}
}

func groupResponse(baseURL string, group corestore.Record) map[string]any {
	return map[string]any{
		"id":                    stringField(group, "okta_id"),
		"created":               group["created_at"],
		"lastUpdated":           group["updated_at"],
		"lastMembershipUpdated": group["updated_at"],
		"objectClass":           []string{"okta:user_group"},
		"type":                  stringField(group, "type"),
		"profile": map[string]any{
			"name":        stringField(group, "name"),
			"description": group["description"],
		},
		"_links": map[string]any{
			"self": map[string]any{
				"href": strings.TrimRight(baseURL, "/") + "/api/v1/groups/" + url.PathEscape(stringField(group, "okta_id")),
			},
		},
	}
}

func appResponse(baseURL string, app corestore.Record) map[string]any {
	return map[string]any{
		"id":          stringField(app, "okta_id"),
		"name":        stringField(app, "name"),
		"label":       stringField(app, "label"),
		"status":      stringField(app, "status"),
		"created":     app["created_at"],
		"lastUpdated": app["updated_at"],
		"signOnMode":  stringField(app, "sign_on_mode"),
		"credentials": mapValue(app["credentials"]),
		"settings":    mapValue(app["settings"]),
		"_links": map[string]any{
			"self": map[string]any{
				"href": strings.TrimRight(baseURL, "/") + "/api/v1/apps/" + url.PathEscape(stringField(app, "okta_id")),
			},
		},
	}
}

func authorizationServerResponse(baseURL string, server corestore.Record) map[string]any {
	serverID := stringField(server, "server_id")
	return map[string]any{
		"id":          serverID,
		"name":        stringField(server, "name"),
		"description": stringField(server, "description"),
		"audiences":   stringSliceValue(server["audiences"]),
		"issuer":      resolveIssuer(baseURL, serverID),
		"status":      stringField(server, "status"),
		"created":     server["created_at"],
		"lastUpdated": server["updated_at"],
		"_links": map[string]any{
			"self": map[string]any{
				"href": strings.TrimRight(baseURL, "/") + "/api/v1/authorizationServers/" + url.PathEscape(serverID),
			},
		},
	}
}

func bodyMap(body map[string]any, key string) map[string]any {
	return mapValue(body[key])
}

func bodyString(body map[string]any, key string) string {
	return strings.TrimSpace(stringValue(body[key]))
}

func bodyStringDefault(body map[string]any, key string, fallback string) string {
	if value := bodyString(body, key); value != "" {
		return value
	}
	return fallback
}

func bodyStringSlice(body map[string]any, key string, fallback []string) []string {
	value, ok := body[key]
	if !ok {
		return append([]string(nil), fallback...)
	}
	items := stringSliceValue(value)
	if len(items) == 0 {
		return append([]string(nil), fallback...)
	}
	return items
}

func writeNoContent(c *corehttp.Context) {
	c.Writer.WriteHeader(http.StatusNoContent)
}
